bundles/http/Citrix/00000019-0000-0000-0000-000000000000/47a6eb95-6f9d-44df-8be7-2322ada5770c.js

```diff
@@ -0,0 +1,864 @@
+}
+
+async function syncAriba(sync, incrementalSync) {
+	console.log('Ariba sync started')
+	let documentsLists = { //creating lists od rq and invc ids
+		requisitions: [],
+		invoices: [],
+		lastChangeId: sync.integrationParameters.lastChangeId,
+		newChangeId: 1
+	}
+	let groups = {
+		ids: [],
+		members: []
+	}
+	if (incrementalSync) {
+		if (sync.context.lastChangeSequenceId > documentsLists.lastChangeId) { // reading stored lastChangeSequenceId for incremental sync
+			documentsLists.lastChangeId = sync.context.lastChangeSequenceId
+		}
+	}
+	else {
+		documentsLists = await fetchApprovablesAndGetDocuments(sync,
+			documentsLists) //geting documents id, this endpoint is used only in full sync hold documents for open approvals infinitely
+	}
+	documentsLists = await fetchChangesGetDocumentsAndGetLastId(sync,
+		documentsLists) //geting documents id, this endpoint is used for full and incr, hold all data only for last 90 days 
+	documentsLists.requisitions = [...new Set(documentsLists.requisitions)] //creating uniq list
+	documentsLists.invoices = [...new Set(documentsLists.invoices)] //creating uniq list
+	groups = await storeRequisition(sync, documentsLists.requisitions, groups) //storing requisitions and resolve group approvables
+	if (documentsLists.newChangeId > documentsLists.lastChangeId) {
+		sync.context.lastChangeSequenceId = documentsLists.newChangeId
+	} //saving  lastChangeSequenceId for incremental sync into context
+	console.log('Max Change ID ' + documentsLists.newChangeId)
+	console.log('Total number of requisitions ' + documentsLists.requisitions.length)
+	console.log('Total number of invoices ' + documentsLists.invoices.length)
+	console.log('Sync finished')
+	return
+}
+
+function fullSync(sync) {
+	return syncAriba(sync, false)
+}
+
+function incrSync(sync) {
+	return syncAriba(sync, true)
+}
+
+async function approveDenyRequisition(sync) {
+	const {
+		approvableId,
+		passwordAdapter,
+		action,
+		user,
+		visibleToSupplier,
+		comment,
+		totalCostAmount,
+		totalCostCurrency
+	} = sync.actionParameters
+	const approveDenyRequest = {
+		'approvableId': approvableId,
+		'comment': comment,
+		'visibleToSupplier': visibleToSupplier
+	}
+	let groups = {
+		ids: [],
+		members: []
+	}
+	let responseData = await sync.serviceClient.fetch('requisitions/' + approvableId + '?realm=' + sync.integrationParameters.realm, {
+		headers: {
+			'apikey': sync.integrationParameters.apiKey
+		}
+	})
+	try {
+		if (responseData.ok) {
+			let json = await responseData.json()
+			if (isObject(json) && Math.abs(json.totalCost.amount - totalCostAmount) < 0.01 && json.totalCost.currency ==
+				totalCostCurrency) { //validation of total costs and currency
+				console.log('Validation ok, total cost and currency match')
+				responseData = await sync.serviceClient.fetch('/operations/rest/requisitions/' + action + '?realm=' + sync.integrationParameters.realm +
+					'&user=' + user + '&passwordadapter=' + passwordAdapter, {
```

Comment:
All parameters which could be with special characters should be encoded with `encodeURIComponent()`.
```suggestion
				responseData = await sync.serviceClient.fetch(`/operations/rest/requisitions/${encodeURIComponent(action)}?realm=${encodeURIComponent(sync.integrationParameters.realm)}&user=${encodeURIComponent(user )}&passwordadapter=${encodeURIComponent(passwordAdapter)}`, {
```

Actually, this should be done on all URLs in `fetch()` calls. But in other places, it is not as urgent as here. Mostly because for example, `groupId` will be probably ok.
But in the case of `sync.integrationParameters.realm` I'm not sure, so I suggest to escape it too.

For more info look into [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent#description)